lib/Controller/UsersController.php

```diff
@@ -469,6 +501,226 @@ public function create($username, $password, array $groups= [], $email='') {
+	 * @param string $userId
+	 * @return null
+	 * @throws UserTokenException
+	 */
+	private function checkPasswordSetToken($token, $userId) {
+		$user = $this->userManager->get($userId);
+
+		$splittedToken = \explode(':', $this->config->getUserValue($userId, 'owncloud', 'lostpassword', null));
+		if (\count($splittedToken) !== 2) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_AND_TIME_INVALID);
+		}
+
+		if ($splittedToken[0] < ($this->timeFactory->getTime() - 60*60*12) ||
+			$user->getLastLogin() > $splittedToken[0]) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_TIME_EXPIRED);
+		}
+
+		if (!\hash_equals($splittedToken[1], $token)) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_MISMACTH);
+		}
+	}
+
+	/**
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @NoSubadminRequired
+	 * @param $userId
+	 */
+	public function resendToken($userId) {
+		$user = $this->userManager->get($userId);
+
+		if ($user === null) {
+			return new TemplateResponse(
+				'core', 'error', [
+				"errors" => [["error" => $this->l10n->t('User: %s does not exist', [$userId])]]
+			],
+				'guest'
+			);
+		}
+
+		if ($user->getEMailAddress() === null) {
+			return new TemplateResponse(
+				'core', 'error', [
+				"errors" => [["error" => $this->l10n->t('Email address not set for: %s', [$userId])]]
+			],
+				'guest'
+			);
+		}
+
+		$token = $this->secureRandom->generate(21,
+			ISecureRandom::CHAR_DIGITS,
+			ISecureRandom::CHAR_LOWER, ISecureRandom::CHAR_UPPER);
+		$this->config->setUserValue($userId, 'owncloud',
+			'lostpassword', $this->timeFactory->getTime() . ':' . $token);
+
+		// data for the mail template
+		$mailData = [
+			'username' => $userId,
+			'url' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.setPasswordForm', ['userId' => $userId, 'token' => $token])
+		];
+
+		$mail = new TemplateResponse('user_management', 'new_user/email-html', $mailData, 'blank');
+		$mailContent = $mail->render();
+
+		$mail = new TemplateResponse('user_management', 'new_user/email-plain_text', $mailData, 'blank');
+		$plainTextMailContent = $mail->render();
+
+		$subject = $this->l10n->t('Your %s account was created', [$this->defaults->getName()]);
+
+		try {
+			$message = $this->mailer->createMessage();
+			$message->setTo([$user->getEMailAddress() => $userId]);
+			$message->setSubject($subject);
+			$message->setHtmlBody($mailContent);
+			$message->setPlainBody($plainTextMailContent);
+			$message->setFrom([$this->fromMailAddress => $this->defaults->getName()]);
+			$this->mailer->send($message);
+		} catch (\Exception $e) {
+			$this->log->error("Can't send new user mail to " . $user->getEMailAddress() . ": " . $e->getMessage(), ['app' => 'settings']);
```

Comment:
just asking, should we show something to the user if the email can't be sent?

Reply:
Done. The user will receive message with the updated pr.